conformance/echo-basic/echo-basic.go

```diff
@@ -201,15 +209,29 @@ func runH2CServer(h2cPort string, errchan chan<- error) {
 }
 
 func echoHandler(w http.ResponseWriter, r *http.Request) {
+	var sni string
+
 	fmt.Printf("Echoing back request made to %s to client (%s)\n", r.RequestURI, r.RemoteAddr)
 
 	// If the request has form ?delay=[:duration] wait for duration
 	// For example, ?delay=10s will cause the response to wait 10s before responding
-	if err := delayResponse(r); err != nil {
+	err := delayResponse(r)
+	if err != nil {
 		processError(w, err, http.StatusInternalServerError)
 		return
 	}
 
+	// If the request was made to URI backendTLS, then get the server name indication and
+	// add it to the RequestAssertions.  It will be echoed back later.
+	if strings.Contains(r.RequestURI, "backendTLS") {
+		sni, err = sniffForSNI(r.RemoteAddr)
```

Comment:
I don't get how this works. Are we supposed to send an HTTP request and then send a TLS request later?

Reply:
If we are just serving this as TLS already you can just do `r.TLS.ServerName` I think

Reply:
The idea is that you send an HTTP or HTTPS request and the BackendTLSPolicy adds what is needed for the backend request to succeed (certificate and SNI).

Reply:
I still need to dig into this a bit deeper, but I think https://github.com/kubernetes-sigs/gateway-api/pull/3212#discussion_r1738642041 is relevant to this question.

Reply:
The logic seems to be: on an HTTP request, on a new TCP listener, accept a **new** connection, get the TLS handshake data, then close the connection.

Should it not be: accept an HTTPS request and report the SNI?

Reply:
If you're saying should we always be testing HTTPs and forget about HTTP, my initial reaction is no because I thought we were saying (though I could understand arguments that this is dubious) that `client -> HTTP -> Gateway -> HTTPs -> Backend` was a valid use case?

Reply:
There are two connections - one terminates at the gateway and one authenticates with the backend.  I was attempting to get the SNI from the second connection here.  Maybe this isn't correct, but just accepting one connection is also not correct.  Advice gladly accepted.

Reply:
This *is* the backend though. We don't want to make a connection to ourself.

Reply:
What we want is 

`Test binary -- (1) TLS 1-->Gateway --(2) TLS 2 --> test app`

Connection 1 is establihsed by the go test. Connection 2 is established by the gateway implementation under test.

The test app does not establish a connection, it accepts one.